lib/trie/proof/verify_test.go

```diff
@@ -0,0 +1,435 @@
+	}
+
+	leafBLarge := node.Node{
+		Key:   []byte{2},
+		Value: longValue,
+	}
+	leafBLargeEncoded := encodeNode(t, leafBLarge)
+	require.Greater(t, len(leafBLargeEncoded), 32)
+
+	leafCLarge := node.Node{
+		Key:   []byte{3},
+		Value: longValue,
+	}
+	leafCLargeEncoded := encodeNode(t, leafCLarge)
+	require.Greater(t, len(leafCLargeEncoded), 32)
+
+	testCases := map[string]struct {
+		encodedProofNodes [][]byte
+		rootHash          []byte
+		expectedTrie      *trie.Trie
+		errWrapped        error
+		errMessage        string
+	}{
+		"no proof node": {
+			errWrapped: ErrEmptyProof,
+			rootHash:   []byte{1},
+			errMessage: "proof slice empty: for Merkle root hash 0x01",
+		},
+		"proof node decoding error": {
+			encodedProofNodes: [][]byte{
+				{1, 2, 3},
+			},
+			rootHash:   []byte{3},
+			errWrapped: node.ErrUnknownNodeType,
+			errMessage: "cannot decode node at index 0: " +
+				"unknown node type: 0 (node encoded is 0x010203)",
+		},
+		"root proof encoding smaller than 32 bytes": {
+			encodedProofNodes: [][]byte{
+				encodeNode(t, leafAShort),
+			},
+			rootHash: blake2bNode(t, leafAShort),
+			expectedTrie: trie.NewTrie(&node.Node{
+				Key:        leafAShort.Key,
+				Value:      leafAShort.Value,
+				Encoding:   leafAShortEncoded,
+				HashDigest: blake2bNode(t, leafAShort),
+			}),
+		},
+		"root proof encoding larger than 32 bytes": {
+			encodedProofNodes: [][]byte{
+				leafBLargeEncoded,
+			},
+			rootHash: blake2bNode(t, leafBLarge),
+			expectedTrie: trie.NewTrie(&node.Node{
+				Key:        leafBLarge.Key,
+				Value:      leafBLarge.Value,
+				Encoding:   leafBLargeEncoded,
+				HashDigest: blake2bNode(t, leafBLarge),
+			}),
+		},
+		"discard unused node": {
+			encodedProofNodes: [][]byte{
+				leafAShortEncoded,
+				leafBLargeEncoded,
+			},
+			rootHash: blake2bNode(t, leafAShort),
+			expectedTrie: trie.NewTrie(&node.Node{
+				Key:        leafAShort.Key,
+				Value:      leafAShort.Value,
+				Encoding:   leafAShortEncoded,
+				HashDigest: blake2bNode(t, leafAShort),
+			}),
+		},
+		"multiple unordered nodes": {
+			encodedProofNodes: [][]byte{
+				leafBLargeEncoded, // chilren 1 and 3
+				encodeNode(t, node.Node{ // root
+					Key: []byte{1},
+					Children: padRightChildren([]*node.Node{
+						&leafAShort, // inlined
+						&leafBLarge, // referenced by Merkle value hash
+						&leafCLarge, // referenced by Merkle value hash
+						&leafBLarge, // referenced by Merkle value hash
+					}),
+				}),
+				leafCLargeEncoded, // children 2
+			},
+			rootHash: blake2bNode(t, node.Node{
+				Key: []byte{1},
+				Children: padRightChildren([]*node.Node{
+					&leafAShort,
+					&leafBLarge,
+					&leafCLarge,
+					&leafBLarge,
+				}),
+			}),
+			expectedTrie: trie.NewTrie(&node.Node{
+				Key:         []byte{1},
+				Descendants: 4,
+				Children: padRightChildren([]*node.Node{
+					{
+						Key:   leafAShort.Key,
+						Value: leafAShort.Value,
+						Dirty: true, // TODO: this should be false
```

Comment:
TODO this should be set to false in the production code.

Reply:
Changed to set it to true in all cases, since the trie isn't used outside of the proof code (nor the database), and also because to be consistent we would have to recursively check child from a decoded node and set their dirty flag to false. that's more, it's more confusing and hits performance a bit.